Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm

```diff
@@ -65,6 +69,122 @@
+    auto aggregator = MainRunLoopCallbackAggregator::create([recordHolder, completionHandler = WTFMove(completionHandler)]() mutable {
+        Vector<Ref<WebExtensionDataRecord>> records;
+        for (auto entry : recordHolder->recordsMap)
+            records.append(entry.value);
+
+        completionHandler(records);
+    });
+
+    auto uniqueIdentifiers = FileSystem::listDirectory(m_configuration->storageDirectory());
+    for (auto& uniqueIdentifier : uniqueIdentifiers) {
+        auto storageDirectory = FileSystem::pathByAppendingComponent(m_configuration->storageDirectory(), uniqueIdentifier);
+        auto stateFilePath = FileSystem::pathByAppendingComponent(storageDirectory, "State.plist"_s);
+
+        String displayName;
+        URL lastBaseURL;
+        if (!WebExtensionContext::readDisplayNameAndLastBaseURLFromState(stateFilePath, displayName, lastBaseURL))
+            continue;
+
+        for (auto type : types) {
+            calculateStorageSizeForDataType(type, storageDirectory, uniqueIdentifier, makeBlockPtr([recordHolder, aggregator, uniqueIdentifier, displayName, type](size_t size) mutable {
+                Ref record = recordHolder->recordsMap.ensure(uniqueIdentifier, [&] {
+                    return WebExtensionDataRecord::create(displayName, uniqueIdentifier);
+                }).iterator->value;
+                record->setSizeOfType(type, size);
+            }));
+        }
+    }
+}
+
+void WebExtensionController::getDataRecord(OptionSet<WebExtensionDataType> types, WebExtensionContext& extensionContext, CompletionHandler<void(RefPtr<WebExtensionDataRecord>)>&& completionHandler)
+{
+    if (!m_configuration->storageIsPersistent() || types.isEmpty()) {
+        completionHandler(nullptr);
+        return;
+    }
+
+    String matchingUniqueIdentifier;
+    String storageDirectory;
+    String displayName;
+
+    auto recordHolder = RecordHolder::create();
+    auto aggregator = MainRunLoopCallbackAggregator::create([recordHolder, completionHandler = WTFMove(completionHandler)]() mutable {
+        completionHandler(recordHolder->recordsMap.takeFirst());
+    });
+
+    auto uniqueIdentifiers = FileSystem::listDirectory(m_configuration->storageDirectory());
+    for (auto& uniqueIdentifier : uniqueIdentifiers) {
+        storageDirectory = FileSystem::pathByAppendingComponent(m_configuration->storageDirectory(), uniqueIdentifier);
+        auto stateFilePath = FileSystem::pathByAppendingComponent(storageDirectory, "State.plist"_s);
+
+        URL lastBaseURL;
+        if (!WebExtensionContext::readDisplayNameAndLastBaseURLFromState(stateFilePath, displayName, lastBaseURL))
+            continue;
+
+        if (this->extensionContext(lastBaseURL)->extension() == extensionContext.extension()) {
+            matchingUniqueIdentifier = uniqueIdentifier;
+            break;
+        }
+    }
+
+    if (!matchingUniqueIdentifier) {
+        completionHandler(nullptr);
+        return;
+    }
```

Comment:
This still assumes the extension context is loaded. While this works, I think we can make this work for unloaded contexts too. The passed in context does not need to be loaded, since we just needs to ask for it's unique identifier.

```c++
auto uniqueIdentifiers = FileSystem::listDirectory(m_configuration->storageDirectory());
if (!uniqueIdentifiers.contains(extensionContext.uniqueIdentifier())) {
    completionHandler(nullptr);
    return;
}

auto uniqueIdentifier = extensionContext.uniqueIdentifier();
auto storageDirectory = FileSystem::pathByAppendingComponent(m_configuration->storageDirectory(), uniqueIdentifier);
auto stateFilePath = FileSystem::pathByAppendingComponent(storageDirectory, "State.plist"_s);

String displayName;
URL lastBaseURL;
if (!WebExtensionContext::readDisplayNameAndLastBaseURLFromState(stateFilePath, displayName, lastBaseURL)) {
    completionHandler(nullptr);
    return;
}
```